homeassistant/components/number/__init__.py

```diff
@@ -0,0 +1,103 @@
+"""Component to allow numeric input for platforms."""
+from datetime import timedelta
+import logging
+from typing import Any, Dict
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from .const import (
+    ATTR_MAX,
+    ATTR_MIN,
+    ATTR_STEP,
+    ATTR_VALUE,
+    DEFAULT_MAX_VALUE,
+    DEFAULT_MIN_VALUE,
+    DEFAULT_STEP,
+    DOMAIN,
+    SERVICE_SET_VALUE,
+)
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Track states and offer events for number entities."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_SET_VALUE,
+        {vol.Required(ATTR_VALUE): vol.Coerce(float)},
+        "async_set_value",
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    return await hass.data[DOMAIN].async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    return await hass.data[DOMAIN].async_unload_entry(entry)
+
+
+class NumberEntity(Entity):
+    """Representation of a Number entity."""
+
+    @property
+    def capability_attributes(self) -> Dict[str, Any]:
+        """Return capability attributes."""
+        data = {
+            ATTR_MIN: self.min_value,
+            ATTR_MAX: self.max_value,
+            ATTR_STEP: self.step,
+        }
+        return data
+
+    @property
+    def min_value(self) -> float:
+        """Return the minimum value."""
+        return DEFAULT_MIN_VALUE
+
+    @property
+    def max_value(self) -> float:
+        """Return the maximum value."""
+        return DEFAULT_MAX_VALUE
+
+    @property
+    def step(self) -> float:
+        """Return the increment/decrement step."""
+        step = DEFAULT_STEP
+        value_range = abs(self.max_value - self.min_value)
```

Comment:
When setting up a `Number` entity, it would be good to check that `self.max_value > self.min_value`. I think the inequality should be strict, not `>=`.

Reply:
I believe this should be a responsibility of the platform to ensure that. Because the platform can overload the min and max properties, they could in theory even be not available when setting up the entity.